### PR TITLE
Make heap size hint available as an env variable

### PR DESCRIPTION
**Pull request by MilesCranmer**
This makes `JULIA_HEAP_SIZE_HINT` the environment variable version of the `--heap-size-hint` command-line flag. Seems like there was interest in https://github.com/JuliaLang/julia/pull/45369#issuecomment-1544204022.

The same syntax is used as for the command-line version with, for example, `2G` => 2 GB and `200M` => 200 MB. 

@oscardssmith want to take a look?

### CODE REVIEW

**Review by oscardssmith — DISMISSED** [review]

**Comment by tecosaur** [comment]
It would be nice if the envvar could also be added to the docs :slightly_smiling_face:

**Comment by MilesCranmer** [comment]
Where should it go? It wasn't clear to me where to put it.

**Comment by MilesCranmer** [comment]
@tecosaur I couldn’t find documentation on `--heap-size-hint` aside from the CLI help message. Where should the env var docs go? I’m happy to add them I just don’t know where.

**Comment by lgoettgens** [comment]
On https://docs.julialang.org/en/v1/manual/environment-variables/, all ENV variables are explained. So I would look for it there to find more information

**Comment by MilesCranmer** [comment]
Done.

**Comment by MilesCranmer** [comment]
Ping. Would be nice to have this.

File: doc/src/manual/environment-variables.md [hunk 1]
```diff
@@ -392,6 +392,23 @@ If set to anything besides `0`, then Julia's thread policy is consistent with
...
+Forces garbage collection if memory usage is higher than the given value. The value may be specified as a number of bytes, optionally in units of:
+
+    - b  (bytes)
+    - k  (kibibytes)
+    - m  (mebibytes)
+    - g  (gibibytes)
+    - t  (tebibytes)
+    - %  (percentage of physical memory)
+
+which are not case sensitive. For example, `JULIA_HEAP_SIZE_HINT=1G` would provide a 1 GB heap size hint to the garbage collector.
```
Comment (IanButterworth) [inline doc/src/manual/environment-variables.md:410]: The units seem a bit inconsistent.. lowercase in the top list, uppercase in the example, then two-letter uppercase in the cli docs
https://github.com/JuliaLang/julia/blob/e1e7763e31d77c2b7c812a2fce4556bfd943f409/src/jloptions.c#L234-L237

Comment (MilesCranmer) [reply doc/src/manual/environment-variables.md:410]: Just for reference – the current method in Julia, for whatever reason, is written as to not be case sensitive and actually only takes the first character – so, e.g., `5GB` would get parsed as `5g`. So technically these are all valid. For example:

```bash
julia --heap-size-hint=1Giga
```

will successfully set a 1 GB heap size hint. This is totally valid (though not documented, I guess?) on the current Julia.

I personally don't have a preference other than I agree consistency is always good. Wdyt? In retrospect I think the parsing should have required a specific case and number of characters. But too late for that now as all Julia back to 1.9 have this quirk.

Comment (MilesCranmer) [reply doc/src/manual/environment-variables.md:410]: The inconsistency is also found in the existing docs. e.g., the current code is
```julia
    "                                               number of bytes, optionally in units of KB, MB, GB,\n"
    "                                               or TB, or as a percentage of physical memory with %.\n\n"
```
while the docs give the single-letter version: https://julialang.org/blog/2023/04/julia-1.9-highlights/#memory_usage_hint_for_the_gc_with_--heap-size-hint

Comment (IanButterworth) [reply doc/src/manual/environment-variables.md:410]: I would've thought matching here would make most sense https://docs.julialang.org/en/v1.12-dev/manual/command-line-interface/ so uppercase 2 chars. (Though I see [1.10](https://docs.julialang.org/en/v1.10/manual/command-line-interface/#cli) does give 'G` as an example.. but better to match current version)

Comment (MilesCranmer) [reply doc/src/manual/environment-variables.md:410]: Ok will do the 2-char version, upper-case, everywhere.

Comment (MilesCranmer) [reply doc/src/manual/environment-variables.md:410]: Hm, I take that back. It looks weird. Note that most unix tools, a single upper-case character is used. For example, `dd`, `rsync`, ulimit` - they all use K for kilobytes, G for gigabytes, M for megabytes. 

Even `ls` will display file sizes with a single character: `ls -l`.

Comment (MilesCranmer) [reply doc/src/manual/environment-variables.md]: Ok I made everything single char in the docs. (Note that it's backwards compatible; it's just the docs were inconsistent before)

**Comment by oscardssmith** [comment]
triage approves.

**Pull request merged**